Cython/Compiler/ExprNodes.py

```diff
@@ -9863,13 +9878,17 @@ def analyse_declarations(self, env):
         self.def_node.is_cyfunction = False
         # Force genexpr signature
         self.def_node.entry.signature = TypeSlots.pyfunction_noargs
+        self.loop.iterator.outer_scope = env
```

Comment:
> Are we sure that the loop or its iterator haven't been replaced at this point, by some preceding transform?

Fairly sure but some more guards to check the assumption might help.

> Could this be set in a more local place, or by the inside node itself? Reaching out seems easier/safer than reaching in.

I'll have a look. I think it's wrapped in quite a dew layers of `GeneratorBodyDefNodes` and similar, so the information about the scope isn't easily available there.

> Also, you seem to be doing the same in `IteratorNode.analyse_declarations()`. Is that the case and intended?

It's in `ComprehensionNode.analyse_declarations` which is intended (for very similar reasons).

Reply:
> Could this be set in a more local place, or by the inside node itself?

I've now done this. It needed a few additional tags on the scopes so that it could identify the right scope to use, but I think it's probably an improvement.